# MATH 210 Introduction to Mathematical Computing

**January 12, 2024**

* Numeric types: integers, floats and complex numbers
* Arithmetic operations
* Examples: Roots and Taylor Series
* Variables

## Numeric types and arithmetic operations

There are 3 main numeric types in Python: integers, floats and complex numbers.

We have the usual arithmetic operations: `+`, `-`, `*`, `/`, `**`.

Add integers:

In [1]:
1 + 2 + 3 + 4 + 5

15

In [2]:
13 - 7

6

In [3]:
5*4*3*2*1

120

In [4]:
7**2

49

In [5]:
100/25

4.0

A float (aka floating point number) is a real number in decimal form.

Division always returns a float even if we are dividing integers where the result is an integer theoretically.

There is (almost) always some rounding error in floating point arithmetic. For example:

In [6]:
0.1 + 0.2

0.30000000000000004

We need to get used to interpreting numbers like $10^{-16}$ in Python as 0.

Use the Python function `type` to determine the type of a value.

In [7]:
type(42)

int

In [8]:
type(42.0)

float

The complex number $\sqrt{-1}$ is entered as `1j` in Python.

In [9]:
2 - 3j

(2-3j)

In [10]:
type(1j)

complex

In [11]:
(-1)**(1/2)

(6.123233995736766e-17+1j)

We can use scientific notation `1e-3` to create floating point numbers.

In [12]:
1e-3

0.001

So what is `(6.123233995736766e-17+1j)`? The real part is `6.123233995736766e-17` which is $6.123233995736766 \times 10^{-17}$ which is basically 0.

There is no $\pi$ built into Python. We need the NumPy package. For now, we can just use:

In [13]:
3.1415926535

3.1415926535

In [14]:
2**0.5

1.4142135623730951

In [15]:
1.4142135623730951**2

2.0000000000000004

Suppose both `type(x)` and `type(y)` are `int`. Determine `type(x ** y)`.

If `y >= 0` then `type(x ** y)` is `int`. If `y < 0` then `type(x ** y)` is `float`. 

In [16]:
2**3

8

In [17]:
5**(-2)

0.04

## Examples: Roots and Taylor Series

The roots of the polynomial $ax^2 + bx + c = 0$ are given by the quadratic formula:

$$
x = \frac{ -b \pm \sqrt{b^2 - 4ac} }{ 2a }
$$

For example, the roots of $x^2 - x - 1 = 0$ are:

In [18]:
(-(-1) + ((-1)**2 - 4*(1)*(-1))**(1/2))/(2*1)

1.618033988749895

In [19]:
(-(-1) - ((-1)**2 - 4*(1)*(-1))**(1/2))/(2*1)

-0.6180339887498949

The Taylor series of $f(x)$ at $x=a$ is

$$
f(x) = \sum_{k=0}^{\infty} \frac{f^{(k)}(a)}{k!} (x - a)^k
$$

The Taylor series of $e^x$ at $a=0$ is

$$
e^x = \sum_{k=0}^{\infty} \frac{1}{k!} (x - 0)^k = \sum_{k=0}^{\infty} \frac{x^k}{k!}
$$

Use partial sum up to $N=5$ to approximate $e$:

$$
e \approx \sum_{k=0}^N \frac{1}{k!}
$$

In [20]:
1/1 + 1/1 + 1/2 + 1/(2*3) + 1/(2*3*4) + 1/(2*3*4*5)

2.7166666666666663

Note that we use parentheses to group operations. We can't use square brackets `[ ]` or `{ }`.

In [21]:
1/[2*3*4]

TypeError: unsupported operand type(s) for /: 'int' and 'list'

In [22]:
1/{2*3*4}

TypeError: unsupported operand type(s) for /: 'int' and 'set'

Compute the sum

$$
\sum_{n=1}^N \frac{1}{n^2}
$$

for $N=10$.

In [23]:
1  + 1/2**2 + 1/3**2 + 1/4**2 + 1/5**2 + 1/6**2 + 1/7**2 + 1/8**2 + 1/9**2 + 1/10**2

1.5497677311665408

## Variables

In [24]:
a = 1
b = -1
c = -1
(-b + (b**2 - 4*a*c)**(1/2))/(2*a)

1.618033988749895

In [25]:
x = -1
x**0/1 + x**1/1 + x**2/2 + x**3/(2*3) + x**4/(2*3*4) + x**5/(2*3*4*5)

0.3666666666666667